rfcs/0108-nixos-containers.md

```diff
@@ -0,0 +1,467 @@
+
+## Imperative management
+
+`systemd` differentiates between "privileged" & "unprivileged" settings. Each privileged (also
+called "trusted") `nspawn` unit lives in `/etc/systemd/nspawn`. Since unprivileged container's
+don't allow bind mounts, these will be out of scope. Additionally, this means that
+`/etc/systemd/nspawn` has to be writable for administrative users and can't be a symlink to
+a store path anymore.
+
+The new implementation is written in Python since it's expected to be more accessible than Perl
+and thus more folks are willing to maintain this code (just as it was the case after porting
+the VM test driver from Perl to Python).
+
+The following features won't be available anymore in the new script:
+
+* Start/Stop operations, logging into containers: this can be entirely done via [`machinectl(1)`](https://www.freedesktop.org/software/systemd/man/machinectl.html).
+* No configuration will be made via CLI flags. Instead, the option set from the
+  NixOS module will be used to declare not only the container's configuration, but also
+  networking. This approach is inspired by [erikarvstedt/extra-container](https://github.com/erikarvstedt/extra-container).
+
+But still, not all features from declarative containers are implemented here, for instance:
+
+* One has to explicitly specify whether to restart/reload a container when updating the config.
+  This is done on purpose to avoid duplicating the logic from `switch-to-configuration.pl` here.
+* IPv6 prefix delegation is turned off because `radvd`'s configuration is declaratively specified
+  when building the host's NixOS.
+
+Examples are in the next chapter.
+
+## Config activation
+
+By default, NixOS has to decide how to activate configuration changes for a container to avoid
+e.g. unnecessary reboots, but `reload`s aren't necessarily sufficient either because e.g.
+new bind mounts require a reboot. The host's `switch-to-configuration.pl` implements it like
+this:
+
+* `systemctl reload systemd-nspawn@container-name.service` runs `switch-to-configuration test`
+  inside the container `container-name` using `nsenter`.
+* When activating a new config on the host, the following things happen:
+  * If the setting `Parameter=` in the container's `.nspawn`-unit is the only thing that has changed,
+    a `reload` will be done. This parameter contains the `init`-script for the container's NixOS
+    and only changes if the container's NixOS config changes.
+  * If anything else changes, `systemd-nspawn@container-name.service` will be scheduled for a restart
+    which effectively reboots the container.
+* This behavior can be turned off completely which means that the container where this is turned
+  off won't be touched at all on `switch-to-configuration`. Additionally, it's possible to always
+  force a `reload` or `restart`. See [Examples & Interactions](#examples-and-interactions) for
+  details.
```

Comment:
This feels like a lot of (maybe unnecessary) complexity, especially the `nsenter` part.

Could we just activate the new configuration inside the container on a reload, and restart the container if any of the mounts in the .nspawn file has changed?

We could provide some option to always force a container restart.

Reply:
why does the container need to be restarted when the port forwards change? How exactly do they affect the container side of things?

Reply:
The `reload` only does a `switch-to-configuration` and thus only alters what happens *inside* the container. To activate the port-forward (via systemd), you can currently only reboot.